File: CS/ElectronicStore.cs
/*
=========================================================
Author      : Razak Nadaf
Roll No.    : 2638
Assignment  : A1.2
Project     : Electronic Store Management System
File Name   : ElectronicStore.cs

Description:
Main electronic store class managing relations between Owner,
Outlet, Employees, Products, Suppliers, Customers, and
Inventory stock. Demonstrates aggregation, composition, and
encapsulation in C#.
=========================================================
*/

using System;
using System.Collections.Generic;

namespace ElectronicsStore
{
    public class ElectronicStore
    {
        private int storeId;
        private string name;
        private string contactNo;
        private Owner owner;
        private Outlet outlet;
        private List<Employee> employees;
        private List<Supplier> suppliers;
        private List<Customer> customers;

        // Constructor connecting all main store components
        public ElectronicStore(int storeId, string name, string contactNo, Owner owner, Outlet outlet)
        {
            this.storeId = storeId;
            this.name = name;
            this.contactNo = contactNo;
            this.owner = owner;
            this.outlet = outlet;
            this.employees = new List<Employee>();
            this.suppliers = new List<Supplier>();
            this.customers = new List<Customer>();
        }

        // Getters and Setters
        public int GetStoreId()
        {
            return storeId;
        }

        public void SetStoreId(int storeId)
        {
            this.storeId = storeId;
        }

        public string GetName()
        {
            return name;
        }

        public void SetName(string name)
        {
            this.name = name;
        }

        public string GetContactNo()
        {
            return contactNo;
        }

        public void SetContactNo(string contactNo)
        {
            this.contactNo = contactNo;
        }

        public Owner GetOwner()
        {
            return owner;
        }

        public void SetOwner(Owner owner)
        {
            this.owner = owner;
        }

        public Outlet GetOutlet()
        {
            return outlet;
        }

        public void SetOutlet(Outlet outlet)
        {
            this.outlet = outlet;
        }

        public Inventory GetInventory()
        {
            return outlet != null ? outlet.GetInventory() : null;
        }

        // --- Product Management (handled via Outlet Inventory) ---
        public void AddProduct(Product product)
        {
            if (outlet != null && outlet.GetInventory() != null)
            {
                outlet.GetInventory().AddProduct(product);
            }
        }

        public bool RemoveProduct(int productId)
        {
            if (outlet != null && outlet.GetInventory() != null)
            {
                return outlet.GetInventory().RemoveProduct(productId);
            }
            return false;
        }

        public Product FindProduct(int productId)
        {
            if (outlet != null && outlet.GetInventory() != null)
            {
                return outlet.GetInventory().FindProduct(productId);
            }
            return null;
        }

        public List<Product> GetProducts()
        {
            if (outlet != null && outlet.GetInventory() != null)
            {
                return outlet.GetInventory().GetProducts();
            }
            return new List<Product>();
        }

        // --- Employee Management ---
        public void AddEmployee(Employee employee)
        {
            employees.Add(employee);
        }

        public List<Employee> GetEmployees()
        {
            return employees;
        }

        // --- Supplier Management ---
        public void AddSupplier(Supplier supplier)
        {
            suppliers.Add(supplier);
            Console.WriteLine(">> Supplier \"" + supplier.GetSupplierName() + "\" added successfully.");
        }

        public List<Supplier> GetSuppliers()
        {
            return suppliers;
        }

        public Supplier FindSupplier(int supplierId)
        {
            foreach (Supplier s in suppliers)
            {
                if (s.GetSupplierId() == supplierId)
                {
                    return s;
                }
            }
            return null;
        }

        // --- Customer Management ---
        public void AddCustomer(Customer customer)
        {
            customers.Add(customer);
            Console.WriteLine(">> Customer \"" + customer.Name + "\" registered successfully.");
        }

        public List<Customer> GetCustomers()
        {
            return customers;
        }

        public Customer FindCustomer(int customerId)
        {
            foreach (Customer c in customers)
            {
                if (c.GetCustomerId() == customerId)
                {
                    return c;
                }
            }
            return null;
        }

        // Display summary of store
        public void DisplayStoreInfo()
        {
            Console.WriteLine("========================================");
            Console.WriteLine("       ELECTRONICS STORE OVERVIEW       ");
            Console.WriteLine("========================================");
            Console.WriteLine("Store ID: " + storeId);
            Console.WriteLine("Store Name: " + name);
            Console.WriteLine("Store Contact: " + contactNo);
            if (owner != null)
            {
                Console.WriteLine("Owner: " + owner.Name + " (Contact: " + owner.ContactNo + ")");
            }
            if (outlet != null)
            {
                Console.WriteLine("Main Outlet: " + outlet.GetName() + " (" + outlet.GetAddress() + ")");
            }
            Console.WriteLine("Total Products: " + GetProducts().Count);
            Console.WriteLine("Total Employees: " + employees.Count);
            Console.WriteLine("Total Suppliers: " + suppliers.Count);
            Console.WriteLine("Total Customers: " + customers.Count);
            Console.WriteLine("========================================");
        }
    }
}

File: Java/ElectronicStore.java
/*
=========================================================
Author      : Razak Nadaf
Roll No.    : 2638
Assignment  : A1.2
Project     : Electronic Store Management System
File Name   : ElectronicStore.java

Description:
Main electronic store class managing relations between Owner,
Outlet, Employees, Products, Suppliers, Customers, and
Inventory stock.
=========================================================
*/

import java.util.ArrayList;

public class ElectronicStore {
    private int storeId;
    private String name;
    private String contactNo;
    private Owner owner;
    private Outlet outlet;
    private ArrayList<Employee> employees;
    private ArrayList<Supplier> suppliers;
    private ArrayList<Customer> customers;

    // Constructor connecting all main store components
    public ElectronicStore(int storeId, String name, String contactNo, Owner owner, Outlet outlet) {
        this.storeId = storeId;
        this.name = name;
        this.contactNo = contactNo;
        this.owner = owner;
        this.outlet = outlet;
        this.employees = new ArrayList<Employee>();
        this.suppliers = new ArrayList<Supplier>();
        this.customers = new ArrayList<Customer>();
    }

    // Getters and Setters
    public int getStoreId() {
        return storeId;
    }

    public void setStoreId(int storeId) {
        this.storeId = storeId;
    }

    public String getName() {
        return name;
    }

    public void setName(String name) {
        this.name = name;
    }

    public String getContactNo() {
        return contactNo;
    }

    public void setContactNo(String contactNo) {
        this.contactNo = contactNo;
    }

    public Owner getOwner() {
        return owner;
    }

    public void setOwner(Owner owner) {
        this.owner = owner;
    }

    public Outlet getOutlet() {
        return outlet;
    }

    public void setOutlet(Outlet outlet) {
        this.outlet = outlet;
    }

    public Inventory getInventory() {
        return outlet != null ? outlet.getInventory() : null;
    }

    // --- Product Management (handled via Outlet Inventory) ---
    public void addProduct(Product product) {
        if (outlet != null && outlet.getInventory() != null) {
            outlet.getInventory().addProduct(product);
        }
    }

    public boolean removeProduct(int productId) {
        if (outlet != null && outlet.getInventory() != null) {
            return outlet.getInventory().removeProduct(productId);
        }
        return false;
    }

    public Product findProduct(int productId) {
        if (outlet != null && outlet.getInventory() != null) {
            return outlet.getInventory().findProduct(productId);
        }
        return null;
    }

    public ArrayList<Product> getProducts() {
        if (outlet != null && outlet.getInventory() != null) {
            return outlet.getInventory().getProducts();
        }
        return new ArrayList<Product>();
    }

    // --- Employee Management ---
    public void addEmployee(Employee employee) {
        employees.add(employee);
    }

    public ArrayList<Employee> getEmployees() {
        return employees;
    }

    // --- Supplier Management ---
    public void addSupplier(Supplier supplier) {
        suppliers.add(supplier);
        System.out.println(">> Supplier \"" + supplier.getSupplierName() + "\" added successfully.");
    }

    public ArrayList<Supplier> getSuppliers() {
        return suppliers;
    }

    public Supplier findSupplier(int supplierId) {
        for (Supplier s : suppliers) {
            if (s.getSupplierId() == supplierId) {
                return s;
            }
        }
        return null;
    }

    // --- Customer Management ---
    public void addCustomer(Customer customer) {
        customers.add(customer);
        System.out.println(">> Customer \"" + customer.getName() + "\" registered successfully.");
    }

    public ArrayList<Customer> getCustomers() {
        return customers;
    }

    public Customer findCustomer(int customerId) {
        for (Customer c : customers) {
            if (c.getCustomerId() == customerId) {
                return c;
            }
        }
        return null;
    }

    // Display summary of store
    public void displayStoreInfo() {
        System.out.println("========================================");
        System.out.println("       ELECTRONICS STORE OVERVIEW       ");
        System.out.println("========================================");
        System.out.println("Store ID: " + storeId);
        System.out.println("Store Name: " + name);
        System.out.println("Store Contact: " + contactNo);
        if (owner != null) {
            System.out.println("Owner: " + owner.getName() + " (Contact: " + owner.getContactNo() + ")");
        }
        if (outlet != null) {
            System.out.println("Main Outlet: " + outlet.getName() + " (" + outlet.getAddress() + ")");
        }
        System.out.println("Total Products: " + getProducts().size());
        System.out.println("Total Employees: " + employees.size());
        System.out.println("Total Suppliers: " + suppliers.size());
        System.out.println("Total Customers: " + customers.size());
        System.out.println("========================================");
    }
}
